### PR TITLE
Reworked plugin support

### PR DESCRIPTION
**Pull request by ash2k**

### CODE REVIEW

File: pkg/apis/smith/v1/types.go [hunk 1]
```diff
@@ -187,6 +187,9 @@ func (bs *BundleStatus) ShortString() string {
 // ResourceName is a reference to another Resource in the same bundle.
 type ResourceName string
 
+// PluginName is a name of a plugin to be invoked.
```
Comment (wryun) [inline pkg/apis/smith/v1/types.go]: what's the advantage of this? (curiosity)

File: cmd/smith/main.go [hunk 2]
```diff
@@ -41,63 +28,9 @@ func runWithContext(ctx context.Context) error {
...
-
-	if *plugins != "" {
-		a.Plugins = strings.Split(*plugins, ",")
-	}
-
-	if pprofAddr != nil && *pprofAddr != "" {
-		go func() {
-			log.Fatalf("pprof server failed: %v", http.ListenAndServe(*pprofAddr, nil))
-		}()
+	if err := a.ParseFlags(flag.CommandLine, os.Args[1:]); err != nil {
```
Comment (wryun) [inline cmd/smith/main.go]: I'd probably push as something that automatically happens on Run (and give CommandLine/Args). Just in the interest of simplicity of interface... I don't see any advantage to forcing the plugin author to call ParseFlags before calling run.

In a similar vein, why not just have an 'a.Run' command which sets up 'a' with a default workers of 2? And if you want a different number of workers, have it passed as a flag?

File: pkg/apis/smith/v1/types.go (hunk 1 above)
Comment (ash2k) [reply pkg/apis/smith/v1/types.go]: Type safety

File: cmd/smith/main.go (hunk 2 above)
Comment (ash2k) [reply cmd/smith/main.go]: 1. Originally `app` package did not depend on flags or configuration files or anything like that. It is/was agnostic of where the configuration comes from. This helps a lot when you want to reuse the code or set up the whole app in an integration test. Imagine setting up an integration test with `.Run()` that wants you to pass a FlagSet into it. Ugly. I was actually thinking to put flag parsing into a separate package and keeping `app` clean.
2. Workers should be a command line flag, yes.

**Pull request merged**